staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go

```diff
@@ -53,13 +53,23 @@ type APIServiceSpec struct {
 	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
 	CABundle []byte
 
-	// Priority controls the ordering of this API group in the overall discovery document that gets served.
-	// Client tools like `kubectl` use this ordering to derive preference, so this ordering mechanism is important.
-	// Values must be between 1 and 1000
-	// The primary sort is based on priority, ordered lowest number to highest (10 before 20).
+	// GroupPriority controls the ordering of the group with respect to other groups. The lowest of all GroupPriorities
+	// for the same group is used.  Must be greater than zero.
+	// The primary sort is based on GroupPriority, ordered lowest number to highest (10 before 20).
+	// The secondary sort is based on the VersionPriority, ordered lowest number to highest.
+	// The tertiary sort is based on the alphabetical comparison of the name of the object.  (v1.bar before v1.foo)
+	// We'd recommend something like: *.k8s.io (except extensions) at 1000, extensions at 2000
+	// PaaSes (OpenShift, Deis) are recommended to be in the 3000s
+	GroupPriority int32
```

Comment:
My 2 cents:

We actually define a `MinGroupPriority` here (let's call it this way to make this explicit!). I would prefer a way that avoids this conceptual complexity, but I agree with the arguments of "self-contained APIServer" and "different servers for group versions" leading to this. So I guess we have to accept this complexity. Neither of the proposed solutions (with the exception of a Groups resources which we don't want for other reasons) solves this, only changes the encoding.

And in this light, splitting `MinGroupPriority` and `VersionPriority` is at least a step forward as it splits a group-local order from a global order. Try to explain to somebody to decrease the priority of a version, but not too much because as it might influence the group order.